One test failure on macOS:

```
=================================== FAILURES ===================================
_____________________________ test_default_queries _____________________________

database = <spack.database.Database object at 0x10b0f5940>

    def test_default_queries(database):
        # Testing a package whose name *doesn't* start with 'lib'
        # to ensure the library has 'lib' prepended to the name
        rec = database.get_record('zmpi')
    
        spec = rec.spec
    
>       libraries = spec['zmpi'].libs

lib/spack/spack/test/database.py:389: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
lib/spack/spack/spec.py:941: in __get__
    value = f()
lib/spack/spack/spec.py:933: in <lambda>
    callbacks_chain.append(lambda: self.default(self, instance, cls))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

descriptor = <spack.spec.ForwardQueryToPackage object at 0x107362280>
spec = zmpi@1.0%gcc@4.5.0 arch=test-debian6-core2 ^fake@1.0%gcc@4.5.0 arch=test-debian6-core2
cls = <class 'llnl.util.lang.Spec'>

    def _libs_default_handler(descriptor, spec, cls):
        """Default handler when looking for the 'libs' attribute.
    
        Tries to search for ``lib{spec.name}`` recursively starting from
        ``spec.prefix``. If ``spec.name`` starts with ``lib``, searches for
        ``{spec.name}`` instead.
    
        Parameters:
            descriptor (ForwardQueryToPackage): descriptor that triggered the call
            spec (Spec): spec that is being queried
            cls (type(spec)): type of spec, to match the signature of the
                descriptor ``__get__`` method
    
        Returns:
            LibraryList: The libraries found
    
        Raises:
            NoLibrariesError: If no libraries are found
        """
    
        # Variable 'name' is passed to function 'find_libraries', which supports
        # glob characters. For example, we have a package with a name 'abc-abc'.
        # Now, we don't know if the original name of the package is 'abc_abc'
        # (and it generates a library 'libabc_abc.so') or 'abc-abc' (and it
        # generates a library 'libabc-abc.so'). So, we tell the function
        # 'find_libraries' to give us anything that matches 'libabc?abc' and it
        # gives us either 'libabc-abc.so' or 'libabc_abc.so' (or an error)
        # depending on which one exists (there is a possibility, of course, to
        # get something like 'libabcXabc.so, but for now we consider this
        # unlikely).
        name = spec.name.replace('-', '?')
    
        # Avoid double 'lib' for packages whose names already start with lib
        if not name.startswith('lib'):
            name = 'lib' + name
    
        # If '+shared' search only for shared library; if '~shared' search only for
        # static library; otherwise, first search for shared and then for static.
        search_shared = [True] if ('+shared' in spec) else \
            ([False] if ('~shared' in spec) else [True, False])
    
        for shared in search_shared:
            libs = fs.find_libraries(
                name, spec.prefix, shared=shared, recursive=True)
            if libs:
                return libs
    
        msg = 'Unable to recursively locate {0} libraries in {1}'
>       raise spack.error.NoLibrariesError(msg.format(spec.name, spec.prefix))
E       spack.error.NoLibrariesError: Unable to recursively locate zmpi libraries in /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pytest-of-runner/pytest-0/mock_store0/test-debian6-core2/gcc-4.5.0/zmpi-1.0-ocvhmmabw3q7ezkozxcd7phdakzocf7l

lib/spack/spack/spec.py:872: NoLibrariesError
= 1 failed, 2587 passed, 128 skipped, 22 xfailed, 1 xpassed in 2033.14 seconds =
```

restarted it to see if it's just flaky

--- a/.flake8
+++ b/.flake8
@@ -29,4 +29,4 @@
 #
 [flake8]
 ignore = E129,E221,E241,E272,E731,W503,W504,F999,N801,N813,N814
-max-line-length = 79
+max-line-length = 88

--- a/lib/spack/llnl/util/lock.py
+++ b/lib/spack/llnl/util/lock.py
@@ -26,6 +26,126 @@ lock_type = {fcntl.LOCK_SH: 'read', fcntl.LOCK_EX: 'write'}
 true_fn = lambda: True
 
 
+class OpenFile(object):
+    """Record for keeping track of open lockfiles (with reference counting).
+
+    There's really only one ``OpenFile`` per inode, per process, but we record the
+    filehandle here as it's the thing we end up using in python code.  You can get
+    the file descriptor from the file handle if needed -- or we could make this track
+    file descriptors as well in the future.
+    """
+    def __init__(self, fh):
+        self.fh = fh
+        self.refs = 0
+
+
+class OpenFileTracker(object):
+    """Track open lockfiles, to minimize number of open file descriptors.
+
+    The ``fcntl`` locks that Spack uses are associated with an inode and a process.
+    This is convenient, because if a process exits, it releases its locks.
+    Unfortunately, this also means that if you close a file, *all* locks associated
+    with that file's inode are released, regardless of whether the process has any
+    other open file descriptors on it.
+
+    Because of this, we need to track open lock files so that we only close them when
+    a process no longer needs them.  We do this by tracking each lockfile by its
+    inode and process id.  This has several nice properties:
+
+    1. Tracking by pid ensures that, if we fork, we don't inadvertently track the parent
+       process's lockfiles. ``fcntl`` locks are not inherited across forks, so we'll
+       just track new lockfiles in the child.
+    2. Tracking by inode ensures that referencs are counted per inode, and that we don't
+       inadvertently close a file whose inode still has open locks.
+    3. Tracking by both pid and inode ensures that we only open lockfiles the minimum
+       number of times necessary for the locks we have.
+
+    Note: as mentioned elsewhere, these locks aren't thread safe -- they're designed to
+    work in Python and assume the GIL.
+    """
+
+    def __init__(self):
+        """Create a new ``OpenFileTracker``."""
+        self._descriptors = {}
+
+    def get_fh(self, path):
+        """Get a filehandle for a lockfile.
+
+        This routine will open writable files for read/write even if you're asking
+        for a shared (read-only) lock. This is so that we can upgrade to an exclusive
+        (write) lock later if requested.
+
+        Arguments:
+          path (str): path to lock file we want a filehandle for
+        """
+        # Open writable files as 'r+' so we can upgrade to write later
+        os_mode, fh_mode = (os.O_RDWR | os.O_CREAT), 'r+'
+
+        pid = os.getpid()
+        open_file = None  # OpenFile object, if there is one
+        stat = None       # stat result for the lockfile, if it exists
+
+        try:
+            # see whether we've seen this inode/pid before
+            stat = os.stat(path)
+            key = (stat.st_ino, pid)
+            open_file = self._descriptors.get(key)
+
+        except OSError as e:
+            if e.errno != errno.ENOENT:  # only handle file not found
+                raise
+
+            # path does not exist -- fail if we won't be able to create it
+            parent = os.path.dirname(path) or '.'
+            if not os.access(parent, os.W_OK):
+                raise CantCreateLockError(path)
+
+        # if there was no already open file, we'll need to open one
+        if not open_file:
+            if stat and not os.access(path, os.W_OK):
+                # we know path exists but not if it's writable. If it's read-only,
+                # only open the file for reading (and fail if we're trying to get
+                # an exclusive (write) lock on it)
+                os_mode, fh_mode = os.O_RDONLY, 'r'
+
+            fd = os.open(path, os_mode)
+            fh = os.fdopen(fd, fh_mode)
+            open_file = OpenFile(fh)
+
+            # if we just created the file, we'll need to get its inode here
+            if not stat:
+                inode = os.fstat(fd).st_ino
+                key = (inode, pid)
+
+            self._descriptors[key] = open_file
+
+        open_file.refs += 1
+        return open_file.fh
+
+    def release_fh(self, path):
+        """Release a filehandle, only closing it if there are no more references."""
+        try:
+            inode = os.stat(path).st_ino
+        except OSError as e:
+            if e.errno != errno.ENOENT:  # only handle file not found
+                raise
+            inode = None  # this will not be in self._descriptors
+
+        key = (inode, os.getpid())
+        open_file = self._descriptors.get(key)
+        assert open_file, "Attempted to close non-existing lock path: %s" % path
+
+        open_file.refs -= 1
+        if not open_file.refs:
+            del self._descriptors[key]
+            open_file.fh.close()
+
+
+#: Open file descriptors for locks in this process. Used to prevent one process
+#: from opening the sam file many times for different byte range locks
+file_tracker = OpenFileTracker()
+
+
 def _attempts_str(wait_time, nattempts):
     # Don't print anything if we succeeded on the first try
     if nattempts <= 1:
@@ -46,7 +166,8 @@ class Lock(object):
     Note that this is for managing contention over resources *between*
     processes and not for managing contention between threads in a process: the
     functions of this object are not thread-safe. A process also must not
-    maintain multiple locks on the same file.
+    maintain multiple locks on the same file (or, more specifically, on
+    overlapping byte ranges in the same file).
     """
 
     def __init__(self, path, start=0, length=0, default_timeout=None,
@@ -151,25 +272,10 @@ class Lock(object):
 
         # Create file and parent directories if they don't exist.
         if self._file is None:
-            parent = self._ensure_parent_directory()
+            self._ensure_parent_directory()
+            self._file = file_tracker.get_fh(self.path)
 
-            # Open writable files as 'r+' so we can upgrade to write later
-            os_mode, fd_mode = (os.O_RDWR | os.O_CREAT), 'r+'
-            if os.path.exists(self.path):
-                if not os.access(self.path, os.W_OK):
-                    if op == fcntl.LOCK_SH:
-                        # can still lock read-only files if we open 'r'
-                        os_mode, fd_mode = os.O_RDONLY, 'r'
-                    else:
-                        raise LockROFileError(self.path)
-
-            elif not os.access(parent, os.W_OK):
-                raise CantCreateLockError(self.path)
-
-            fd = os.open(self.path, os_mode)
-            self._file = os.fdopen(fd, fd_mode)
-
-        elif op == fcntl.LOCK_EX and self._file.mode == 'r':
+        if op == fcntl.LOCK_EX and self._file.mode == 'r':
             # Attempt to upgrade to write lock w/a read-only file.
             # If the file were writable, we'd have opened it 'r+'
             raise LockROFileError(self.path)
@@ -282,7 +388,8 @@ class Lock(object):
         """
         fcntl.lockf(self._file, fcntl.LOCK_UN,
                     self._length, self._start, os.SEEK_SET)
-        self._file.close()
+
+        file_tracker.release_fh(self.path)
         self._file = None
         self._reads = 0
         self._writes = 0

--- a/lib/spack/spack/environment.py
+++ b/lib/spack/spack/environment.py
@@ -1420,6 +1420,13 @@ class Environment(object):
         # DB read transaction to reduce time spent in this case.
         specs_to_install = self.uninstalled_specs()
 
+        # ensure specs already installed are marked explicit
+        all_specs = [cs for _, cs in self.concretized_specs()]
+        specs_installed = [s for s in all_specs if s.package.installed]
+        with spack.store.db.write_transaction():  # do all in one transaction
+            for spec in specs_installed:
+                spack.store.db.update_explicit(spec, True)
+
         if not specs_to_install:
             tty.msg('All of the packages are already installed')
             return

--- a/lib/spack/spack/test/cmd/env.py
+++ b/lib/spack/spack/test/cmd/env.py
@@ -176,6 +176,27 @@ def test_env_install_single_spec(install_mockery, mock_fetch):
     assert e.specs_by_hash[e.concretized_order[0]].name == 'cmake-client'
 
 
+def test_env_roots_marked_explicit(install_mockery, mock_fetch):
+    install = SpackCommand('install')
+    install('dependent-install')
+
+    # Check one explicit, one implicit install
+    dependent = spack.store.db.query(explicit=True)
+    dependency = spack.store.db.query(explicit=False)
+    assert len(dependent) == 1
+    assert len(dependency) == 1
+
+    env('create', 'test')
+    with ev.read('test') as e:
+        # make implicit install a root of the env
+        e.add(dependency[0].name)
+        e.concretize()
+        e.install_all()
+
+    explicit = spack.store.db.query(explicit=True)
+    assert len(explicit) == 2
+
+
 def test_env_modifications_error_on_activate(
         install_mockery, mock_fetch, monkeypatch, capfd):
     env('create', 'test')

